### ISSUE TITLE
PSC failed with Nvidia GPU

### ISSUE THREAD

**Issue by wdsun1008** [AMDESE/AMDSEV#174]
I'm trying to use SNP with Nvidia GPU passthrough. The virt-install command I'm using was:
```
--host-device=pci_0000_25_00_0 \
--qemu-commandline="-cpu EPYC-v4 -machine memory-encryption=sev0,vmport=off -object memory-backend-memfd-private,id=ram1,size=${ITEMS[2]}M,share=true -object sev-snp-guest,id=sev0,cbitpos=51,reduced-phys-bits=1,${DIGEST},auth-key-enabled=on,discard=both,kernel-hashes=on -machine memory-backend=ram1,kvm-type=protected" 
```
Then I got guest kernel PSC failed error and the VM hang forever:
```
SNP: PSC failed ret=0 exit_info_2=100000
```
VM without GPU passthrough and without SNP are all good. Then I changed discard=both to discard=none, problem solved.
Is there any docs about UPM? I'm curious about the differences between these discard modes. I've tried to do some demos before, using private mem fd and libOS to build a purely software TEE, but I'm not sure about the relationship between UPM and SNP now.

**Comment by mdroth** [comment]
Do you see any messages from the host kernel or QEMU when the failure occurs?

**Comment by wdsun1008** [comment]
> Do you see any messages from the host kernel or QEMU when the failure occurs?

I didn't see any messages, but after using discard=none, I encountered another problem where both nvidia-smi, TensorFlow, and PyTorch can recognize the GPU, but when I use CUDA, it freezes and the CPU usage is at 100% while the GPU usage is at 0%. 
I noticed an issue from [Kata Containers #78](https://github.com/confidential-containers/confidential-containers/issues/78) that mentioned UPM and GPU are currently not compatible. Therefore, I rolled back to the 5.19 rc6 kernel and snp-v3 qemu and started the VM in the same way as before. However, when I used nvidia-smi, the host kernel crashed, and I had to restart the physical machine. I'm not sure what the reason is and was wondering if @zvonkok could give me some advice.

**Comment by wdsun1008** [comment]
@zvonkok Thanks so much! Is there any possible to run non-CC GPU with CVM?

**Comment by zvonkok** [comment]
@wdsun1008 
> DMAs are fully untrusted in a TEE, hence anything will fail with non-CC GPUs.

**Comment by wdsun1008** [comment]
> @wdsun1008
> 
> > DMAs are fully untrusted in a TEE, hence anything will fail with non-CC GPUs.

OK, I got it. I think it would be useful to combine CVM with non-CC GPUs. It may not be entirely safe, but it could be considered as an option to make CVM more widely used.

**Comment by fffmonkeyking-new** [comment]
> > [@wdsun1008](https://github.com/wdsun1008)
> > > DMAs are fully untrusted in a TEE, hence anything will fail with non-CC GPUs.
> 
> OK, I got it. I think it would be useful to combine CVM with non-CC GPUs. It may not be entirely safe, but it could be considered as an option to make CVM more widely used.

What modifications need to be made to enable CVM to use non-CC GPUs for computing?

**Comment by wdsun1008** [comment]
> > > [@wdsun1008](https://github.com/wdsun1008)
> > > > DMAs are fully untrusted in a TEE, hence anything will fail with non-CC GPUs.
> > 
> > 
> > OK, I got it. I think it would be useful to combine CVM with non-CC GPUs. It may not be entirely safe, but it could be considered as an option to make CVM more widely used.
> 
> What modifications need to be made to enable CVM to use non-CC GPUs for computing?

[sev-snp-gpu](https://github.com/JaewonHur/sev-snp-gpu)
Two known issues should be noted: 
1. This implementation is not secure; it's merely a workaround when GPU hardware does not support CC. 
2. Under SNP CVM, running out of GPU memory (OOM) might lead to a host crash, whereas in non-SNP CVM, it would only result in the process being killed.

**Comment by fffmonkeyking-new** [comment]
> > > > [@wdsun1008](https://github.com/wdsun1008)
> > > > > DMAs are fully untrusted in a TEE, hence anything will fail with non-CC GPUs.
> > > 
> > > 
> > > OK, I got it. I think it would be useful to combine CVM with non-CC GPUs. It may not be entirely safe, but it could be considered as an option to make CVM more widely used.
> > 
> > 
> > What modifications need to be made to enable CVM to use non-CC GPUs for computing?
> 
> [sev-snp-gpu](https://github.com/JaewonHur/sev-snp-gpu) Two known issues should be noted:
> 
> 1. This implementation is not secure; it's merely a workaround when GPU hardware does not support CC.
: On machines without CC GPUs, we can accept this approach.

> 2. Under SNP CVM, running out of GPU memory (OOM) might lead to a host crash, whereas in non-SNP CVM, it would only result in the process being killed.
：Under SNP CVM，Why does out of GPU memory (OOM) occur?
：Under TDX CVM，Is it GPU OOM that causes the process to be killed?
：Does non-SNP CVM refer to non-CVM or TDX CVM?

**Comment by wdsun1008** [comment]
> > > > > [@wdsun1008](https://github.com/wdsun1008)
> > > > > > DMAs are fully untrusted in a TEE, hence anything will fail with non-CC GPUs.
> > > > 
> > > > 
> > > > OK, I got it. I think it would be useful to combine CVM with non-CC GPUs. It may not be entirely safe, but it could be considered as an option to make CVM more widely used.
> > > 
> > > 
> > > What modifications need to be made to enable CVM to use non-CC GPUs for computing?
> > 
> > 
> > [sev-snp-gpu](https://github.com/JaewonHur/sev-snp-gpu) Two known issues should be noted:
> > 
> > 1. This implementation is not secure; it's merely a workaround when GPU hardware does not support CC.
> >    : On machines without CC GPUs, we can accept this approach.
> 
> > 2. Under SNP CVM, running out of GPU memory (OOM) might lead to a host crash, whereas in non-SNP CVM, it would only result in the process being killed.
> >    ：Under SNP CVM，Why does out of GPU memory (OOM) occur?
> >    ：Under TDX CVM，Is it GPU OOM that causes the process to be killed?
> >    ：Does non-SNP CVM refer to non-CVM or TDX CVM?

1. In my case, OOM (Out of Memory) occurred when the batch size for LLM inference service was too large. 
2. TDX CVM requires targeted modifications, especially since all current CVM host kernels (6.0+) utilize mechanisms like UPM. This repository is an SEV implementation based on kernel version 5.19-rc6.
3. Non-SNP refers to regular SEV CVM, and the reason for host crashes with SNP may be due to SNP's stricter memory access checks.

**Comment by fffmonkeyking-new** [comment]
“In my case, OOM (Out of Memory) occurred when the batch size for LLM inference service was too large.”
：Is the LLM inference service you run in SNP CVM with None-CC-GPU based on the CUDA framework?

**Comment by wdsun1008** [comment]
> “In my case, OOM (Out of Memory) occurred when the batch size for LLM inference service was too large.” ：Is the LLM inference service you run in SNP CVM with None-CC-GPU based on the CUDA framework?

Yes, I'm not an AI/LLM engineer, just use open source LLM for test, llama/chatglm/qwen.

**Comment by fffmonkeyking-new** [comment]
@wdsun1008 

Have you ever encountered the problem described in the following link in your past practice?
https://github.com/intel/tdx-linux/issues/8
